test/unit/org/apache/cassandra/db/guardrails/GuardrailPageSizeTest.java

```diff
@@ -0,0 +1,126 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.db.guardrails;
+
+import java.util.Collections;
+
+import org.junit.Before;
+import org.junit.Test;
+
+import org.apache.cassandra.config.DatabaseDescriptor;
+import org.apache.cassandra.cql3.CQLStatement;
+import org.apache.cassandra.cql3.QueryOptions;
+import org.apache.cassandra.cql3.QueryProcessor;
+import org.apache.cassandra.db.ConsistencyLevel;
+import org.apache.cassandra.service.ClientState;
+import org.apache.cassandra.service.QueryState;
+import org.apache.cassandra.transport.ProtocolVersion;
+
+import static java.lang.String.format;
+
+/**
+ * Tests the guardrail for the page size, {@link Guardrails#pageSize}.
+ */
+public class GuardrailPageSizeTest extends ThresholdTester
+{
+    private static final int PAGE_SIZE_WARN_THRESHOLD = 5;
+    private static final int PAGE_SIZE_ABORT_THRESHOLD = 10;
+
+    public GuardrailPageSizeTest()
+    {
+        super(PAGE_SIZE_WARN_THRESHOLD,
+              PAGE_SIZE_ABORT_THRESHOLD,
+              DatabaseDescriptor.getGuardrailsConfig().getPageSize(),
+              Guardrails::setPageSizeThreshold,
+              Guardrails::getPageSizeWarnThreshold,
+              Guardrails::getPageSizeAbortThreshold);
+    }
+
+    @Before
+    public void setupTest()
+    {
+        createTable("CREATE TABLE IF NOT EXISTS %s (k INT, c INT, v TEXT, PRIMARY KEY(k, c))");
+    }
+
+    @Test
+    public void testSelectStatementAgainstPageSize() throws Throwable
+    {
+        assertPagingValid("SELECT * FROM %s", 3);
+        assertPagingValid("SELECT * FROM %s", PAGE_SIZE_WARN_THRESHOLD);
+
+        assertPagingWarns("SELECT * FROM %s", 6, format("Query for table %s with page size %s exceeds warning threshold of %s.", currentTable(), 6, PAGE_SIZE_WARN_THRESHOLD));
+        assertPagingWarns("SELECT * FROM %s", PAGE_SIZE_ABORT_THRESHOLD, format("Query for table %s with page size %s exceeds warning threshold of %s.", currentTable(), PAGE_SIZE_ABORT_THRESHOLD, PAGE_SIZE_WARN_THRESHOLD));
+
+        assertPagingAborts("SELECT * FROM %s", 11, format("Aborting %s table query, page size %s exceeds abort threshold of %s.", currentTable(), 11, PAGE_SIZE_ABORT_THRESHOLD));
+    }
+
+    @Test
+    public void testIgnoreInternalClients() throws Throwable
+    {
+        assertPagingIgnored("SELECT * FROM %s", PAGE_SIZE_WARN_THRESHOLD);
+        assertPagingIgnored("SELECT * FROM %s", PAGE_SIZE_ABORT_THRESHOLD);
```

Comment:
These queries won't trigger the intended guardrail because they use exactly the threshold value, so we can't verify that the superuser and the internal user are allowed to skip the guardrail (although the abort threshold check would trigger the warn threshold). I think we should use a higher value:
```suggestion
        assertPagingIgnored("SELECT * FROM %s", PAGE_SIZE_WARN_THRESHOLD + 1);
        assertPagingIgnored("SELECT * FROM %s", PAGE_SIZE_ABORT_THRESHOLD + 1);
```

Reply:
sorry for that - I should spot this issue :/ fixed.